### PR TITLE
feat(composer+git): Add support for custom git repositories with composer

### LINKED ISSUE #3811: Support vcs type for private packages in composer.json
**What would you like Renovate to be able to do?**
We started using the Renovatebot docker container on our company's private Gitlab server. It work's great with our projects that use NPM. But our most important project is using PHP and we developed a lot of smaller packages that we use. 

Each package has it's own Gitlab repository. We tried a few solutions and what worked best is to release a version on git, then specify the package as following in `composer.json`:

```
    "require": {
		"companyname/php-toolbox": "1.2.0",
    },
    "repositories": [
        {
            "type": "vcs",
            "url": "git@gitlab.companyname.io:companyname/php-toolbox.git"
        },
		...
	]
```

However, when scanning, Renovatebot prints out: 

```
 INFO: Dependency lookup failure: not found (repository=companyname/our-php-project, dependency=companyname/php-toolbox)
 INFO: Failed to look up dependency companyname/php-toolbox (repository=companyname/our-php-project, packageFile=composer.json, dependency=companyname/php-toolbox)
 INFO: Unsupported Packagist registry URL (repository=companyname/our-php-project)
       "regUrl": {"type": "vcs", "url": "git@gitlab.companyname.companyname/php-toolbox.git"}
```

By looking at the code in https://github.com/renovatebot/renovate/blob/master/lib/datasource/packagist/index.js#L250-L263 it looks like `vcs` is not a supported type.

**Describe the solution you'd like**
Add support for `vcs` php packages in composer.json if possible. We already pass an access token for Gitlab (with api scope) as `RENOVATE_TOKEN`.

More info:
 
* https://getcomposer.org/doc/05-repositories.md#loading-a-package-from-a-vcs-repository
* https://barryvanveen.nl/blog/55-installing-a-private-package-with-composer

**Describe alternatives you've considered**
For now, we will just manually update those packages without renovatebot automation.

**Additional context**
/

### PR DESCRIPTION
**Pull request by Tanuel**
This is my first time contributing to an open source project, so bear with me ;)

_(Should each commit get its own PR? Let me know and i will split it up if possible)_

Relates to #3811 

# Why?
- Instead of waiting for an update to handle  #3811, I decided to help out an awesome project.
- I am currently working on integrating renovatebot to work on our self-hosted gitlab installation, which is not reachable from outside our intranet
- Composer repositories are referenced as `"url": "git@internal-gitlab:project.git"`
- The following changes will allow usage of self-hosted Renovatebot to work with composer git repositories in a self-hosted GitLab
- Tested by running Renovatebot in an alpine based docker image against our own GitLab instance

# This PR adds three features:

## Custom git repositories with Composer
- repository types git and vcs will be interpreted as git repositories and use git-tags as datasource
    - _Is there a high risk that other vcs types besides git are used?_
- repositories will be read from composer.json and can be defined in config.composer.repositories
- kinda relies on the git-tags changes below, as ssh repository urls can not be resolved with isomorphic-git
- might require additional authentication `composer config -g http-basic.{domain} {username} {password}`

## Allow git-tags datasource to read from HostRules
- With this change, renovate can clone repositories that require authentication
```javascript
// config.js
hostRules: [{
    hostType: 'git',
    url: 'example.com',
    username: 'myUser',
    // use either token or password
    password: 'myPassword',
    // token: 'myToken', 
}]
```

## Semi-support for git@domain:path repositories
- Transform ssh+git urls (git@github.com:project) to https urls, since isomorphic-git does not support 
    - _Possibly change this to actually use ssh via native git call?_
- Why?
    - Composer allows using ssh for custom repositories. However, isomorphic-git does not support this.
    - This will allow renovate to clone git from https urls, but composer can still update via ssh

## What still needs to be done?
While i tried to improve documentation, i am not quite sure what else needs to be updated.

Besides docs, i guess tests need to be updated/improved as well

## Future Scope
- Handle other repository types like github, gitlab

### CODE REVIEW

**Comment by rarkins** [comment]
This is great! I haven’t reviewed the code yet - just on mobile - but wanted to say that I’d be very happy to replace isomorphic git with regular git if that makes it easier for you. I was even considering doing it already because it seems a shame to add a large dependency for a single function call. 

If doing so, check if using simple-git instead of git directly is better for you. It handles the child process and stdout nicely.

**Comment by rarkins** [comment]
And also thank you for your contribution so far. I’ll help you get it to a mergeable state and add you to the contributors list too!

File: lib/manager/composer/extract.js [hunk 1]
```diff
@@ -23,11 +68,28 @@ async function extractPackageFile(content, fileName) {
...
+          if (repositories[depName]) {
+            switch (repositories[depName].type) {
+              case 'vcs':
+              case 'git':
+                datasource = 'gitTags';
+                lookupName = repositories[depName].url;
+                break;
+              default:
+                break;
+            }
```
Comment (Tanuel) [inline lib/manager/composer/extract.js]: I put this in a switch case to easier support other dependency types in the future

**Comment by rarkins** [comment]
Suggestion: 
- a separate PR that simply refactors out isomorphic git and replaces it with simple-git (if possible) or raw git. 
- then, the next PR builds on that to support authentication
- we keep merging from master into this PR until only the Composer part remains

My only doubt is whether we can support ssh-based git *and* authentication via host rules. Can the auth be embedded within the URL, or there some other approach?

**Comment by Tanuel** [comment]
> - a separate PR that simply refactors out isomorphic git and replaces it with simple-git (if possible) or raw git.
> - then, the next PR builds on that to support authentication

This sounds great. As far as i see simple-git uses native git calls? Isomorphic-git itself doesn't require a native git installation.

> My only doubt is whether we can support ssh-based git _and_ authentication via host rules. Can the auth be embedded within the URL, or there some other approach?

In my "solution" in this PR, i simply transformed ssh urls to https urls (no support for ssh). To solve this, one could instead use ssh for repositories that match the pattern `/^git@(.+?):(.+)/` and host rule authentication in other cases. 

In the ssh case, a ssh key needs to be provided somehow. Could be via config, envVar or .ssh/id_rsa. My current approach using a docker image is to add a ssh id_rsa and known_hosts, this would allow composer to update and install via git ssh.

```docker
FROM alpine:3.10 as intermediate
ARG SSH_PRIVATE_KEY
RUN apk add --update --no-cache openssh
RUN mkdir /root/.ssh
RUN echo "${SSH_PRIVATE_KEY}" > /root/.ssh/id_rsa
RUN ssh-keyscan our_gitlab >> /root/.ssh/known_hosts
RUN chmod 600 /root/.ssh/id_rsa

FROM alpine:3.10
COPY --from=intermediate /root/.ssh /root/.ssh
# install yarn, git, curl, openssh, php, composer
```

Https based git repositories would fail to [composer update](https://github.com/renovatebot/renovate/blob/master/lib/manager/composer/artifacts.js#L131) until i added `RUN composer config -g http-basic.{domain} {username} {password}` to my image, mainly because i did not quite understand how to implement it via config.

**Comment by Tanuel** [comment]
Closed in favor of #4055

**Pull request closed without merging**